pymc3/gp/gp.py

```diff
@@ -787,3 +791,151 @@ def conditional(self, name, Xnew, pred_noise=False, given=None, **kwargs):
         return pm.MvNormal(name, mu=mu, chol=chol, shape=shape, **kwargs)
+
+
+@conditioned_vars(["Xs", "y", "noise"])
+class MarginalKron:
+    R"""
+    Marginal Gaussian process with Kronecker structure.
+
+    The `gp.MarginalKron` class is an implementation of the sum of a GP
+    prior and additive noise.  It has `marginal_likelihood`, `conditional`
+    and `predict` methods.  This GP implementation can be used to
+    implement regression on data that is normally distributed.  For more
+    information on the `prior` and `conditional` methods, see their docstrings.
+
+    Parameters
+    ----------
+    cov_func : None, 2D array, or instance of Covariance
+        The covariance function.  Defaults to zero.
+    mean_func : None, instance of Mean
+        The mean function.  Defaults to zero.
+
+    Examples
+    --------
+    .. code:: python
+
+        # A one dimensional column vector of inputs.
+        X = np.linspace(0, 1, 10)[:, None]
+
+        with pm.Model() as model:
+            # Specify the covariance function.
+            cov_func = pm.gp.cov.ExpQuad(1, ls=0.1)
+
+            # Specify the GP.  The default mean function is `Zero`.
+            gp = pm.gp.Marginal(cov_func=cov_func)
+
+            # Place a GP prior over the function f.
+            sigma = pm.HalfCauchy("sigma", beta=3)
+            y_ = gp.marginal_likelihood("y", X=X, y=y, noise=sigma)
+
+        ...
+
+        # After fitting or sampling, specify the distribution
+        # at new points with .conditional
+        Xnew = np.linspace(-1, 2, 50)[:, None]
+
+        with model:
+            fcond = gp.conditional("fcond", Xnew=Xnew)
+    """
+
+    def __init__(self, mean_func=Zero(), cov_funcs=[Constant(0.0)]):
+        self.mean_func = mean_func
+        try:
+            self.cov_funcs = list(cov_funcs)
+            if len(self.cov_funcs) == 1:
+                self.cov_funcs = cycle(self.cov_funcs)
+        except TypeError:
+            self.cov_funcs = cycle([cov_funcs])
+        self.total_cov_func = reduce(mul, self.cov_funcs)
+
+    def _build_marginal_likelihood(self, Xs):
+        self.X = cartesian(*Xs)
+        mu = self.mean_func(self.X)
+        covs = [f(X) for f, X in zip_longest(self.cov_funcs, Xs)]
+        return mu, covs
+
+    def marginal_likelihood(self, name, Xs, y, noise, is_observed=True, **kwargs):
+        """
+        Returns the marginal likelihood distribution, given the input
+        locations `X` and the data `y`.
+        """
+        # if not isinstance(noise, Covariance):
+        #     noise = pm.gp.cov.WhiteNoise(noise)
+        mu, covs = self._build_marginal_likelihood(Xs)
+        self.Xs = Xs
+        self.y = y
+        self.noise = noise
+        if is_observed:
+            return pm.KroneckerNormal(name, mu=mu, covs=covs, noise=noise,
+                                      observed=y, **kwargs)
+        else:
+            # shape = infer_shape(Xs, kwargs.pop("shape", None))
```

Comment:
may need to have shape be a list of shapes for each X?

Reply:
I think the product of the shapes would do, what do you think?